src/stats/manager_impl.go

```diff
@@ -0,0 +1,119 @@
+// @return new stats.
+func (this *ManagerImpl) NewStats(key string) RateLimitStats {
+	ret := RateLimitStats{}
+	logger.Debugf("Creating stats for key: '%s'", key)
+	ret.Key = key
+	ret.TotalHits = this.rlStatsScope.NewCounter(key + ".total_hits")
+	ret.OverLimit = this.rlStatsScope.NewCounter(key + ".over_limit")
+	ret.NearLimit = this.rlStatsScope.NewCounter(key + ".near_limit")
+	ret.OverLimitWithLocalCache = this.rlStatsScope.NewCounter(key + ".over_limit_with_local_cache")
+	ret.WithinLimit = this.rlStatsScope.NewCounter(key + ".within_limit")
+	return ret
+}
+
+type ShouldRateLimitLegacyStats struct {
+	ReqConversionError   gostats.Counter
+	RespConversionError  gostats.Counter
+	ShouldRateLimitError gostats.Counter
+}
+
+func (this *ManagerImpl) NewShouldRateLimitLegacyStats() ShouldRateLimitLegacyStats {
+	return ShouldRateLimitLegacyStats{
+		ReqConversionError:   this.legacyStatsScope.NewCounter("req_conversion_error"),
+		RespConversionError:  this.legacyStatsScope.NewCounter("resp_conversion_error"),
+		ShouldRateLimitError: this.legacyStatsScope.NewCounter("should_rate_limit_error"),
+	}
+}
+
+type ShouldRateLimitStats struct {
+	RedisError   gostats.Counter
+	ServiceError gostats.Counter
+}
+
+func (this *ManagerImpl) NewShouldRateLimitStats() ShouldRateLimitStats {
+	s := this.serviceStatsScope.Scope("call.should_rate_limit")
+	ret := ShouldRateLimitStats{}
+	ret.RedisError = s.NewCounter("redis_error")
+	ret.ServiceError = s.NewCounter("service_error")
+	return ret
+}
+
+type ServiceStats struct {
+	ConfigLoadSuccess gostats.Counter
+	ConfigLoadError   gostats.Counter
+	ShouldRateLimit   ShouldRateLimitStats
+}
+
+func (this *ManagerImpl) NewServiceStats() ServiceStats {
+	ret := ServiceStats{}
+	ret.ConfigLoadSuccess = this.serviceStatsScope.NewCounter("config_load_success")
+	ret.ConfigLoadError = this.serviceStatsScope.NewCounter("config_load_error")
+	ret.ShouldRateLimit = this.NewShouldRateLimitStats()
+	return ret
+}
+
+func DescriptorKey(domain string, descriptor *pb_struct.RateLimitDescriptor) string {
+	rateLimitKey := ""
+	for _, entry := range descriptor.Entries {
+		if rateLimitKey != "" {
+			rateLimitKey += "."
+		}
+		rateLimitKey += entry.Key
+		if entry.Value != "" {
+			rateLimitKey += "_" + entry.Value
+		}
+	}
+	return domain + "." + rateLimitKey
+}
+
+// Stats for an individual rate limit config entry.
+//todo: Ideally the gostats package fields should be unexported
+//	the inner value could be interacted with via getters such as rlStats.TotalHits() uint64
+//	This ensures that setters such as Inc() and Add() can only be managed by ManagerImpl.
+type RateLimitStats struct {
+	Key                     string
+	TotalHits               gostats.Counter
+	OverLimit               gostats.Counter
+	NearLimit               gostats.Counter
+	OverLimitWithLocalCache gostats.Counter
+	WithinLimit             gostats.Counter
+}
+
+func (this RateLimitStats) String() string {
```

Comment:
this may be just me, but it's not obvious that to get `Key` field from `RateLimitStats` type , one should call `String()` method. I would rather call this method `Key()` or `GetKey()`.